### PR TITLE
Dispatch only

### PR DESCRIPTION
**Pull request by taschini**
This is an experiment steering Reg towards an API that is somewhat closer in spirit to that of ``singledispatch`` or ``Peak-rules``.

Each generic function has its own ``Registry``, but there is no need to access it directly. The revised ``usage.rst`` shows in detail how to define generic functions, but a quick example should give an idea:

```python
  @reg.dispatch('item')
  def size(item):
      raise NotImplementedError

  size.register(document_size, item=Document)
  size.register(folder_size, item=Folder)
  size.register(image_size, item=Image)
  size.register(file_size, item=File)
```

Since every generic function always know which lookup to use (its own), the implicit lookup machinery is no longer needed and is removed.

### CODE REVIEW

File: reg/reify.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    Use as a method decorator.  It operates almost exactly like the
+    Python ``@property`` decorator, but it puts the result of the
+    method it decorates into the instance dict after the first call,
+    effectively replacing the function it decorates with an instance
+    variable.  It is, in Python parlance, a non-data descriptor. An
+    example:
+
+    .. testcode::
+
+      from morepath import reify
```
Comment (henri-hulski) [inline reg/reify.py]: @taschini  This should be `from reg import reify` in this case.

**Comment by henri-hulski** [comment]
@taschini 
Don't know about the internals, but the API you're proposing looks really nice and clean.  👍

**Comment by faassen** [comment]
I wish we'd discussed this a bit more before you put in all this work, though I also tremendously appreciate how much you delved into Reg internals. Here are the things I like:

* I like the ``.register`` API and it's one I've considered before, but there are also problems with it,
which I'll discuss below.

* If you have one registry per function there's no more reason to register the dispatch function itself anymore in the registry API, which would simplify it, though you haven't made that simplification yet (you pass in ``self``)

* implicit is gone which makes the API much simpler. Calling one dispatch function from another one is easy and still entirely explicit.

Now as to the problems:

* one reason I haven't gone with a ``register`` method is because Morepath cannot afford to execute registrations immediately. It needs to execute predicate and predicate fallback registrations before any actual functions are registered. Dectate's directive executing deferment and directive ordering facility does this stuff. Of course you could still use this kind of API underneath, but it's going to be hidden under Morepath directives anyway in our primary use case.

* another reason however I haven't gone with a ``register`` method underneath is that we need to support multiple registries; Morepath actively uses one registry per app. This approach doesn't appear to support this use case -- you cannot switch the lookup.

* another reason for a unified single registry and lookup is that Reg can implement a unified cache for these. With lots of small caches you can't implement a LRU cache for all lookups in an app, like Morepath does.

These reasons are why in the branch I've been working on I've gone for a different approach:

* dispatch functions become dispatch methods; they need to be associated with a a class.

* they try to access the ``lookup`` attribute of the class they're on -- so if they're on the app class, they get that lookup. This means lookup switching is the same as app switching.

This covers all use cases I just discussed:

* no more implicit lookup, no more ``lookup`` argument. This simplifies the publisher a bit and ``request.view`` quite a bit.

* different registry per app.

* different lookup per app through the ``lookup`` attribute.

* cache policy for a whole bunch of dispatch methods.

* you can easily call one dispatch method from another as you have access to the ``self`` argument, though in Morepath you proposed (and me too) we call that the ``app`` argument and make it optional.

**Comment by taschini** [comment]
> one reason I haven't gone with a ``register`` method is because Morepath cannot afford to execute registrations immediately. It needs to execute predicate and predicate fallback registrations before any actual functions are registered. Dectate's directive executing deferment and directive ordering facility does this stuff. Of course you could still use this kind of API underneath, but it's going to be hidden under Morepath directives anyway in our primary use case.

What you are describing affects ``register_function`` whether you invoke it directly or via ``Dispatch.register``, so I don't think it is really relevant to this discussion.

> another reason however I haven't gone with a ``register`` method underneath is that we need to support multiple registries; Morepath actively uses one registry per app. This approach doesn't appear to support this use case -- you cannot switch the lookup.

Now, ``RegRegistry`` instances are currently used for two purposes:

1. Provide the machinery for dispatching generic functions on their arguments.
2. Provide a binding scope for functions: e.g., so that in Morepath 0.14  ``generic.permits`` invokes the  ``generic.settings`` defined for the same app, not that for a different app.

These are better provided by two different classes, with different multiplicity constraints:

1. One dispatch registry for each generic function.
2. One scope registry for each scope (app class in the case of Morepath).

In this pull request ``RegRegistry`` is used only for dispatching (hence the name ``dispatch-only`` for the branch), while scoping can be provided by a different class outside of Reg (for instance the ``ProxyRegistry`` class in morepath/morepath#476).

Perhaps I should point out that if I pushed this branch is because I already have a rough prototype for Morepath using this variant of Reg.

> another reason for a unified single registry and lookup is that Reg can implement a unified cache for these. With lots of small caches you can't implement a LRU cache for all lookups in an app, like Morepath does.

Separating the dispatch registries from scope registries immediately shows that they have different caching requirements: only dispatch registries can profit from a cache.  Any reasonable implementation of a scope registry (like the ``ProxyRegistry`` mentioned above) cannot be made any faster by a cache.

File: reg/reify.py (hunk 1 above)
Comment (taschini) [inline reg/reify.py]: Well spotted, thanks.

**Comment by faassen** [comment]
There are two reasons why a ``register`` method on a dispatch function have problems:

* ordering issues: predicate directives need to be executed before dispatch functions are registered. You can get around this by saying predicates are only used for views and no other functions, but that's not what you're saying so I don't understand what you're saying.

* multiple registries. The same dispatch function needs to have multiple functions registered. If there's only a top-level function that you register on, I don't see how this works.

Your idea on how to separate dispatch from having different implementations available is interesting. I completely missed your Morepath pull request with the proxy registry I think.

So to understand what you are proposing:

* Reg functions contain their own registries. They don't care about application switching anymore.

* Morepath (or Reg) implements a special scope registry that allow you to attach functions to it. All application-dependent function calls go through it. All application-dependent function registrations go through it too.

* Instead of doing ``reg_registry.register_function`` in Morepath, you do ``scope_registry.foo.register``.

That is an interesting strategy, and I'm sorry we didn't manage to coordinate our discussions better so far. Here are a few questions:

* making a normal function in scope in Morepath is just assigning to it: ``scope_registry.foo = implementation``. But a dispatch function would need to be made available before you can assign to it. How would this work? You'd need a different version of the same dispatch function on each scope, with its own registry. Note that with the work I've been doing this is handled by declaring methods on the App class.

* the approach still won't have a universal cache policy per scope; you'd have many small caches. This as opposed to a cache per lookup in the current approach. (While I agree that you cannot speed up scope lookups with a cache, this wasn't the intent of the caching system in the first place so I'm not sure what you're getting at there.)

* how does a scoped function call another scoped function? I.e. how does a scoped function get access to the scope? (the scope may be the application object itself). I really would like to understand your approach here. I think you'd end up with a similar approach to mine, where you can make a registration of a dispatch function that takes the scope as its first argument, and during configuration time you detect whether that argument is required. But note that with a method-dispatch approach this is really just Python method calls.

Meanwhile, I already have a completely adjusted Morepath on my branch. I also am working on a general solution to the ``settings`` access use case, and I have a good handle to solve the scope (app) access use case in a reasonable way, with the modifications described here:

https://github.com/morepath/morepath/issues/448#issuecomment-236566465

I am attracted to your notion of separating concerns of scope and dispatch, but I am wondering whether this approach, once you take care of all the use cases, is going to be all that different from what the dispatch method system of Reg is doing now. I think one argument in favor of separate scopes is getting rid of the ``__get__`` magic I have to do now to make the dispatch methods work, but I am wondering what the impact of explicit scope passing is going to be.

Note that I think multi-scope handling should be a feature of Reg in the end that Morepath makes use of, but if we can cleanly separate it in Reg from the handling of dispatches I'm definitely interested.

**Comment by taschini** [comment]
> ordering issues: predicate directives need to be executed before dispatch functions are registered. You can get around this by saying predicates are only used for views and no other functions, but that's not what you're saying so I don't understand what you're saying.

What I'm saying is that this is not a problem.  Let's look again at the implementation of ``Dispatch.register``:

```python
def register(self, value, **key_dict):
    self.registry.register_function(self, value, **key_dict)
```

If you need to something on the ``RegRegistry`` before you can invoke ``register_function(f,…)`` you can pretty much do the same before you invoke ``f.register(…)``.  You can also use [``f.register_external_predicates``](https://github.com/morepath/reg/pull/37/files#diff-3d8637ebf4356b95d0e5693994cd8f88R56) and ``f.register_dispatch_predicates``.

> multiple registries. The same dispatch function needs to have multiple functions registered. If there's only a top-level function that you register on, I don't see how this works.

Suppose that you have two Morepath apps, AppA, and AppB, each with its own scoping registry (say ``AppA.config.scope`` and ``AppB.config.scope``) and you want to have a custom ``dump_json`` for lists in AppA and a custom ``dump_json`` for dicts in AppB.  Dectate will register the first with ``AppA.config.scope.dump_json.register`` and the second with ``AppB.config.scope.dump_json.register``.  There is no top-level function.

**NB** ``App.config.scope.dump_json`` is not a method of the scoping registry. It is a generic function (in the sense of this pull request, i.e.m an instance ``reg.Dispatch``) that happens to be an attribute of a scoping registry.

> Reg functions contain their own registries. They don't care about application switching anymore.

Correct.

> Morepath (or Reg) implements a special scope registry that allow you to attach functions to it. All application-dependent function calls go through it. All application-dependent function registrations go through it too.

Correct.  Note that the implementation of such a scoping registry is a little more than a collection of defaults plus an ``__init__``

> Instead of doing ``reg_registry.register_function`` in Morepath, you do ``scope_registry.foo.register``.

That's right.

> making a normal function in scope in Morepath is just assigning to it: ``scope_registry.foo = implementation``. But a dispatch function would need to be made available before you can assign to it. How would this work? You'd need a different version of the same dispatch function on each scope, with its own registry. Note that with the work I've been doing this is handled by declaring methods on the App class.

You instantiate them in ``scope_registry.__init__``.

> the approach still won't have a universal cache policy per scope; you'd have many small caches. This as opposed to a cache per lookup in the current approach. (While I agree that you cannot speed up scope lookups with a cache, this wasn't the intent of the caching system in the first place so I'm not sure what you're getting at there.)

That's right. You need a different caching strategy.

> how does a scoped function call another scoped function? I.e. how does a scoped function get access to the scope? (the scope may be the application object itself). I really would like to understand your approach here. I think you'd end up with a similar approach to mine, where you can make a registration of a dispatch function that takes the scope as its first argument, and during configuration time you detect whether that argument is required. But note that with a method-dispatch approach this is really just Python method calls.

It's just a matter of function signature. For instance, in the case of ``permits`` (morepath/morepath#475) that would be:

```python
permits(app, identity, obj, permission)
```

**Comment by faassen** [comment]
So the idea is to create new dispatch function objects when you instantiate the scope registry. Cool, that would work.

Concerning the app argument, using a different function signature is the approach I take too. But consider this: a registered dispatch function ``a`` that needs to call another one, ``b``, in the same scope:

```
def a(app, arg):
    app.config.proxy.b(app, arg)
```

You need to say ``app`` twice here, once to get the right scope, and again because you need to pass it in explicitly. That's annoying. You can make it going away by implementing a little helper function for ``b``:

```
class App:
    def b(self, arg):
          return self.config.proxy.b(self, arg)
```

and now you in ``a`` you can write:

```
def a(app, arg):
    app.b(arg)
```

Much nicer. In the method-based dispatch this is done automatically, as ``self`` is the application object. So you can write the latter right away without writing additional wrapping code.

This is the reason I went with method-based dispatch.

I believe in the method-dispatch version of Reg we already separate the notion of scope from the notion of dispatch. Scope is handled by the fact that the dispatch functions are methods of the app class, and that each app class has its own reg registry and thus a lookup. The reg registry itself or lookup knows nothing about scope; there is no implicit lookup anymore. Bound methods *are* like instances of dispatch functions on different app classes, with syntactical support from Python.

**Comment by taschini** [comment]
> You can make it going away by implementing a little helper function for b:

> ```python
class App:
    def b(self, arg):
          return self.config.proxy.b(self, arg)
```

> and now you in a you can write:
> ```python
def a(app, arg):
    app.b(arg)
```

Yes, that's the next step.  And now the interesting question is: what part of the code is in the best place to add that little helper method ``App.b``: Reg or Dectate?

**Comment by faassen** [comment]
Reg on the dispatchmethods branch already has dispatch methods that do this for you, including letting you define a default implementation. It uses Python's method mechanism so make sure the right scope is in use. The scope is contained in the ``lookup`` attribute on the app instance (as opposed to a ``config.scope_registry``).

How would this helper-method approach here be any better than the approach I implemented? We seem to have two parallel efforts to accomplish the same thing, and I get the impression you think the way I accomplished this is suboptimal, but I don't quite understand the reasoning yet and it seems the solutions are pretty convergent. I really would like to understand what your problems are with the dispatchmethods approach, so we can make it better or dump it. We can discuss this here but perhaps a this point chat is more efficient.

**Comment by taschini** [comment]
Superseded by #40.

**Pull request closed without merging**